pkg/chunk/cache/fifo_cache.go

```diff
@@ -279,38 +281,9 @@ func (c *FifoCache) Get(ctx context.Context, key string) (interface{}, bool) {
 	return nil, false
 }
 
-func sizeOf(i interface{}) int {
-	switch v := i.(type) {
-	case string:
-		return len(v)
-	case []int8:
-		return len(v)
-	case []uint8:
-		return len(v)
-	case []int32:
-		return len(v) * 4
-	case []uint32:
-		return len(v) * 4
-	case []float32:
-		return len(v) * 4
-	case []int64:
-		return len(v) * 8
-	case []uint64:
-		return len(v) * 8
-	case []float64:
-		return len(v) * 8
-	// next 2 cases are machine dependent
-	case []int:
-		if l := len(v); l > 0 {
-			return int(unsafe.Sizeof(v[0])) * l
-		}
-		return 0
-	case []uint:
-		if l := len(v); l > 0 {
-			return int(unsafe.Sizeof(v[0])) * l
-		}
-		return 0
-	default:
-		return int(unsafe.Sizeof(i))
-	}
+func sizeOf(item *cacheEntry) int {
+	return int(unsafe.Sizeof(*item)) + // size of cacheEntry
+		cap(item.value) + // size of value
+		(2 * len(item.key)) + // counting key twice: in the cacheEntry and in the map
+		int(unsafe.Sizeof(&list.Element{})) // size of the pointer to an element in the map
```

Comment:
`sizeOf` is little more tricky that I'd like.

For example, I don't really think string length should be counted twice, because it's only string descriptor that's copied, but not the actual string data.

Also `cap(item.value)` is not the full story, as subslice of large slice will report smaller `cap`, but still keep entire large slice referenced.

`int(unsafe.Sizeof(&list.Element{}))` – why counting just a pointer and not full Element struct?

Personally I would simplify it to `len(item.key) + cap(item.value)`, since it's just a best-effort guess, ignoring maintainance structs.

Reply:
Thanks, I updated `sizeOf`.